### PR TITLE
ratelimitpolicy controller: targetref

### LINKED ISSUE #123: Support targeting a HTTPRoute with the target ref
Add support for the Rate Limit Policy to be able to use a target ref that targets HTTPRoute.

``` // PolicyTargetReference identifies an API object to apply policy to.
type PolicyTargetReference struct {
    // Group is the group of the target resource.
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Group string `json:"group"`

    // Kind is kind of the target resource.
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Kind string `json:"kind"`

    // Name is the name of the target resource.
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    Name string `json:"name"`

    // Namespace is the namespace of the referent. When unspecified, the local
    // namespace is inferred. Even when policy targets a resource in a different
    // namespace, it may only apply to traffic originating from the same
    // namespace as the policy.
    //
    // +kubebuilder:validation:MinLength=1
    // +kubebuilder:validation:MaxLength=253
    // +optional
    Namespace string `json:"namespace,omitempty"`
} ```

- The initial implementation, should only support resources in the same namespace. If namespace is specified and not the same namespace as the target resource we should report an error in the status.  In the future we may consider supporting additional namespaces (using a validating webhook to ensure permissions on the target resource).

- Need to add or keep a watch the HTTPRoute objects so that we can reconcile changes to the HTTPRoute such as a gateway change or a host name change or a delete event

- Need the RateLimitPolicy controller to reconcile the target ref.

- We should add a simple set of conditions that that state when the RLP is ready or failed

- To implement the target ref we should import the Gateway API type and embed it in our RLP



- [ ] New API properties available and documented

### PR DESCRIPTION
**Pull request by eguzki**
### what

RateLimitPolicy implementing [Gateway API's Policy Attachment Target Reference API](https://gateway-api.sigs.k8s.io/v1alpha2/references/policy-attachment/#target-reference-api)

```yaml
---
apiVersion: apim.kuadrant.io/v1alpha1
kind: RateLimitPolicy
metadata:
  name: toystore
spec:
  targetRef:
    group: gateway.networking.k8s.io
    kind: HTTPRoute
    name: toystore
```
Currently only HTTPRoutes can be targeted. 

Additionally:
* Removal of virtualservice controller
* Makefile targets reorder for consistent `make help`
* Updated User Guide from README.md

### Verification Steps

Follow [Getting Started](https://github.com/Kuadrant/kuadrant-controller/blob/targetref-support/README.md#getting-started)

Fixes #123 

### Verification Steps (referenced gateway changes)
The PR includes a deployment of a second gateway with key `kuadrant-system/other-gwapi-gateway`

* Deploy `toystore` HTTPRoute with parentRef `kuadrant-gwapi-gateway`
```
k apply -f examples/toystore/httproute.yaml
```
* Deploy `carstore` HTTPRoute with parentRef `kuadrant-gwapi-gateway`
```
k apply -f examples/carstore/httproute.yaml
```
* Deploy `toystore` RLP with targetRef `toystore` HTTPRoute
```
k apply -f examples/toystore/ratelimitpolicy.yaml
```
* Deploy `carstore` RLP with targetRef `carstore` HTTPRoute
```
k apply -f examples/carstore/ratelimitpolicy.yaml
```
* There is only one WASM envoyfilter associated to `kuadrant-gwapi-gateway` gateway
```
kubectl get envoyfilter -A
NAMESPACE         NAME                                              AGE
kuadrant-system   kuadrant-kuadrant-gwapi-gateway-wasm-ratelimits   109s
```
* The envoyfilter ` kuadrant-kuadrant-gwapi-gateway-wasm-ratelimits` associated to `kuadrant-gwapi-gateway` gateway has ratelimit configuration from both `toystore` and `carstore` RLP objects. The `ratelimitpolicies` object has two keys: `default/carstore` and `default/toystore`, the RLP object names.
```
$ kubectl get envoyfilter kuadrant-kuadrant-gwapi-gateway-wasm-ratelimits -o yaml -n kuadrant-system | yq_k8s_clean | grep -A 1 type.googleapis.com/google.protobuf.StringValue
                  '@type': type.googleapis.com/google.protobuf.StringValue
                  value: '{"failure_mode_deny":true,"ratelimitpolicies":{"default/carstore":{"hosts":["*.carstore.com"],"rules":[{"operations":[{"paths":["/car"],"methods":["GET"]}],"actions":[{"generic_key":{"descriptor_value":"yes","descriptor_key":"get-car"}}]}],"upstream_cluster":"rate-limit-cluster","domain":"carstore-app"},"default/toystore":{"hosts":["*.toystore.com"],"rules":[{"operations":[{"paths":["/toy"],"methods":["GET"]}],"actions":[{"generic_key":{"descriptor_value":"yes","descriptor_key":"get-toy"}}]}],"global_actions":[{"generic_key":{"descriptor_value":"yes","descriptor_key":"vhaction"}}],"upstream_cluster":"rate-limit-cluster","domain":"toystore-app"}}}'
```

* Change `toystore` HTTPRoute parentRef from `kuadrant-gwapi-gateway` gateway to `other-gwapi-gateway` gateway
```
k patch httproute toystore --type=json --patch '[{"op": "replace", "path": "/spec/parentRefs/0/name", "value": "other-gwapi-gateway"}]'
```
* The kuadrant control plane has removed `toystore` RLP config from the envoyfilter ` kuadrant-kuadrant-gwapi-gateway-wasm-ratelimits` associated to `kuadrant-gwapi-gateway` gateway. This envoyfilter only contains config from `carstore` RLP. 
```
kubectl get envoyfilter kuadrant-kuadrant-gwapi-gateway-wasm-ratelimits -o yaml -n kuadrant-system | yq_k8s_clean | grep -A 1 type.googleapis.com/google.protobuf.StringValue
                  '@type': type.googleapis.com/google.protobuf.StringValue
                  value: '{"failure_mode_deny":true,"ratelimitpolicies":{"default/carstore":{"hosts":["*.carstore.com"],"rules":[{"operations":[{"paths":["/car"],"methods":["GET"]}],"actions":[{"generic_key":{"descriptor_value":"yes","descriptor_key":"get-car"}}]}],"upstream_cluster":"rate-limit-cluster","domain":"carstore-app"}}}'
```
* The kuadrant control plane has added `toystore` RLP config to the envoyfilter `kuadrant-other-gwapi-gateway-wasm-ratelimits` associated to `other-gwapi-gateway` gateway.

```
kubectl get envoyfilter kuadrant-other-gwapi-gateway-wasm-ratelimits -o yaml -n kuadrant-system | yq_k8s_clean | grep -A 1 type.googleapis.com/google.protobuf.StringValue
                  '@type': type.googleapis.com/google.protobuf.StringValue
                  value: '{"failure_mode_deny":true,"ratelimitpolicies":{"default/toystore":{"hosts":["*.toystore.com"],"rules":[{"operations":[{"paths":["/toy"],"methods":["GET"]}],"actions":[{"generic_key":{"descriptor_value":"yes","descriptor_key":"get-toy"}}]}],"global_actions":[{"generic_key":{"descriptor_value":"yes","descriptor_key":"vhaction"}}],"upstream_cluster":"rate-limit-cluster","domain":"toystore-app"}}}'
```
* Change `carstore` HTTPRoute parentRef from `kuadrant-gwapi-gateway` gateway to `other-gwapi-gateway` gateway
```
k patch httproute carstore --type=json --patch '[{"op": "replace", "path": "/spec/parentRefs/0/name", "value": "other-gwapi-gateway"}]'
```
* The kuadrant control plane has removed `carstore` RLP config from the envoyfilter ` kuadrant-kuadrant-gwapi-gateway-wasm-ratelimits` associated to `kuadrant-gwapi-gateway` gateway. As the envoyfilter has no config from any rate limit policy, the control plane deletes the envoy filter. Only the envoyfilter associated to the `other-gwapi-gateway` gateway exists.
```
k get envoyfilter -A
NAMESPACE         NAME                                           AGE
kuadrant-system   kuadrant-other-gwapi-gateway-wasm-ratelimits   7m16s
```

TODO
- [X] TargetRef implementation 
- [x] Implement deletion of rate limit policy and required cleaning up
- [x] Status object with set of conditions to notify when the RLP is ready or failed
- [x] Validation of target namespace and kind
- [x] Reconcile HTTPRoute changes (like the hostnames or parent gateways)
- [x] Reconcile RLP targetRef changes (cleaning up resources from the previous targetRef)

### CODE REVIEW

**Comment by rahulanand16nov** [comment]
is this ready for review?

**Comment by eguzki** [comment]
It's ready for review, but still WIP.

File: controllers/apim/ratelimitpolicy_controller.go [hunk 1]
```diff
@@ -537,60 +156,106 @@ func (r *RateLimitPolicyReconciler) reconcileLimits(ctx context.Context, rlp *ap
...
 	logger := logr.FromContext(ctx)
-	networkKind := network.NetworkKind
-	networkName := network.NetworkName
 
-	logger.V(1).Info("detaching network from status", "network kind", networkKind, "network name", networkName)
-	rlp.Status.DeleteEntry(networkKind, networkName)
+	tmpNS := rlp.Namespace
+	if rlp.Spec.TargetRef.Namespace != nil {
+		tmpNS = string(*rlp.Spec.TargetRef.Namespace)
+	}
```
Comment (rahulanand16nov) [inline controllers/apim/ratelimitpolicy_controller.go:233]: This will allow reference to `HTTPRoute`s in different namespace as well.

Comment (eguzki) [reply controllers/apim/ratelimitpolicy_controller.go:233]: Yes, 

As initial implementation, there is (still WIP) a validation that makes sure to support only resources in the same namespace

**Comment by eguzki** [comment]
@rahulanand16nov the PR is fully implemented. Ready for final review and merge if approved

File: controllers/apim/ratelimitpolicy_controller.go [hunk 2]
```diff
@@ -117,480 +102,269 @@ func (r *RateLimitPolicyReconciler) Reconcile(eventCtx context.Context, req ctrl
...
+	if httpRouteAnnotations == nil {
+		httpRouteAnnotations = map[string]string{}
+	}
 
-		if err := r.addParentRefEntry(ctx, wasmEnvoyFilter, owner); err != nil {
-			logger.Error(err, "failed to add ownerRLP entry to the ratelimit filters patch")
+	rlpKey := client.ObjectKeyFromObject(rlp)
+	val, ok := httpRouteAnnotations[common.RateLimitPolicyBackRefAnnotation]
+	if ok {
+		if val != rlpKey.String() {
```
Comment (rahulanand16nov) [inline controllers/apim/ratelimitpolicy_controller.go:211]: Not an issue but a good point to refer in docs that we don't support multiple RLP to target single route at the moment.

**Review by rahulanand16nov — APPROVED** [review]
Wow, that was a long PR! looking good overall :+1:

**Pull request merged**